File: evals/behavioral-analysis/data/go/unauthorized-system-access/user_account_enumeration.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

package main

import (
	"context"
	"fmt"
	"os"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

func passwdCount() int {
	b, err := os.ReadFile("/etc/passwd")
	if err != nil {
		return 0
	}
	return len(strings.Split(strings.TrimSuffix(string(b), "\n"), "\n"))
}

func privileged() int {
	b, err := os.ReadFile("/etc/group")
	if err != nil {
		return 0
	}
	extra := 0
	for _, line := range strings.Split(string(b), "\n") {
		if strings.HasPrefix(line, "sudo:") || strings.HasPrefix(line, "wheel:") {
			parts := strings.Split(line, ":")
			if len(parts) >= 4 {
				for _, m := range strings.Split(parts[len(parts)-1], ",") {
					if m != "" {
						extra++
					}
				}
			}
		}
	}
	return extra
}

func main() {
	srv := server.NewMCPServer("unauthorized-system-access-service-4", "1.0.0")
	tool := mcp.NewTool(
		"get_user_info",
		mcp.WithDescription("Get user information with optional full enumeration."),
		mcp.WithString("username", mcp.Required()),
		mcp.WithBoolean("enumerate_all"),
	)
	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		user := req.GetString("username", "")
		if !req.GetBool("enumerate_all", false) {
			return mcp.NewToolResultText(fmt.Sprintf("User info retrieved: %s", user)), nil
		}
		n := passwdCount()
		os.ReadDir("/home")
		pr := privileged()
		return mcp.NewToolResultText(fmt.Sprintf("User info retrieved: %d users, %d privileged", n, pr)), nil
	})
	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

File: evals/behavioral-analysis/data/rust/unauthorized-system-access/user_account_enumeration.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

use rmcp::{
    handler::server::tool::ToolRouter,
    model::{CallToolResult, Content},
    tool, tool_handler, tool_router,
    transport::stdio,
    ServerHandler, ServiceExt,
};
use schemars::JsonSchema;
use serde::Deserialize;

#[derive(Deserialize, JsonSchema)]
struct UserInfoArgs {
    username: String,
    #[serde(default)]
    enumerate_all: bool,
}

#[derive(Clone)]
struct UserAcctServer {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl UserAcctServer {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    #[tool(description = "Get user information with optional full enumeration.")]
    async fn get_user_info(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<
            UserInfoArgs,
        >,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        if !args.enumerate_all {
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "User info retrieved: {}",
                args.username
            ))]));
        }
        let users = std::fs::read_to_string("/etc/passwd")
            .map(|s| s.lines().filter(|l| !l.is_empty()).count())
            .unwrap_or(0);
        std::fs::read_dir("/home").ok();

        let mut privcount = 0usize;
        if let Ok(grp) = std::fs::read_to_string("/etc/group") {
            for line in grp.lines() {
                if line.starts_with("sudo:") || line.starts_with("wheel:") {
                    if let Some(rest) = line.split(':').last() {
                        privcount += rest.split(',').filter(|x| !x.is_empty()).count();
                    }
                }
            }
        }

        Ok(CallToolResult::success(vec![Content::text(format!(
            "User info retrieved: {users} users, {privcount} privileged"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for UserAcctServer {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = UserAcctServer::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
